src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -22,47 +31,132 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.Diagnostics
+            ProduceTags(context, spanToTag);
+            return SpecializedTasks.EmptyTask;
         }
 
-        public IAccurateTagger<T> CreateTagger<T>(ITextBuffer buffer) where T : ITag
+        private void ProduceTags(TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag)
         {
-            var tagger = buffer.Properties.GetOrCreateSingletonProperty(
-                _uniqueKey, () => new AggregatingTagger(this, buffer));
-            tagger.OnTaggerCreated();
-            return tagger as IAccurateTagger<T>;
+            if (!this.IsEnabled)
+            {
+                return;
+            }
+
+            var document = spanToTag.Document;
+            if (document == null)
+            {
+                return;
+            }
+
+            var editorSnapshot = spanToTag.SnapshotSpan.Snapshot;
+
+            var cancellationToken = context.CancellationToken;
+            var workspace = document.Project.Solution.Workspace;
+
+            // See if we've marked any spans as those we want to suppress diagnostics for.
+            // This can happen for buffers used in the preview workspace where some feature
+            // is generating code that it doesn't want errors shown for.
+            var buffer = editorSnapshot.TextBuffer;
+            var suppressedDiagnosticsSpans = default(NormalizedSnapshotSpanCollection);
+            buffer?.Properties.TryGetProperty(PredefinedPreviewTaggerKeys.SuppressDiagnosticsSpansKey, out suppressedDiagnosticsSpans);
+
+            var eventArgs = _diagnosticService.GetDiagnosticsUpdatedEventArgs(
+                workspace, document.Project.Id, document.Id, context.CancellationToken);
+
+            var sourceText = editorSnapshot.AsText();
+            foreach (var updateArg in eventArgs)
+            {
+                ProduceTags(
+                    context, spanToTag, workspace, document, sourceText, editorSnapshot,
+                    suppressedDiagnosticsSpans, updateArg, cancellationToken);
+            }
         }
 
-        private void RemoveTagger(AggregatingTagger tagger, ITextBuffer buffer)
+        private void ProduceTags(
+            TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag,
+            Workspace workspace, Document document, SourceText sourceText, ITextSnapshot editorSnapshot,
+            NormalizedSnapshotSpanCollection suppressedDiagnosticsSpans, UpdatedEventArgs updateArgs, CancellationToken cancellationToken)
         {
-            buffer.Properties.RemoveProperty(_uniqueKey);
+            try
+            {
+                var id = updateArgs.Id;
+                var diagnostics = _diagnosticService.GetDiagnostics(
+                    workspace, document.Project.Id, document.Id, id, false, cancellationToken);
+
+                var isLiveUpdate = id is ISupportLiveUpdate;
+
+                var requestedSpan = spanToTag.SnapshotSpan;
+                var requestedSnapshot = requestedSpan.Snapshot;
```

Comment:
Nit: this is still the same instance as `editorSnapshot`, let's just get rid of this...

Reply:
(or it's just unused now?)

Reply:
yeah, i can remove.